File: server-rails-single/app/models/scopes/task_scope.rb
# frozen_string_literal: true

module Scopes
  # TaskScope — single-tenant ownership isolation (inherited via parent Project).
  #
  # A Task has no user_id of its own; it is owned by whoever owns its Project.
  # We therefore constrain to tasks whose project belongs to the current user.
  #
  # Auto-discovered by the HasAutoScope concern via naming convention. With no
  # request user (console/seeds) `user` is nil and no filter is applied.
  class TaskScope < Rhino::ResourceScope
    def apply(relation)
      return relation unless user

      relation.where(project_id: Project.unscoped.where(user_id: user.id).select(:id))
    end
  end
end

File: server-nestjs-single/src/scopes/TaskScope.ts
import type { RhinoScope } from '@rhino-dev/rhino-nestjs';

/**
 * TaskScope — single-tenant ownership isolation (inherited via parent Project).
 *
 * A Task has no userId of its own; it is owned by whoever owns its Project. We
 * therefore constrain to tasks whose project belongs to the current user, via a
 * Prisma relation filter.
 */
export class TaskScope implements RhinoScope {
  apply(where: Record<string, any>, ctx: { user?: any }) {
    if (ctx.user?.id) {
      return { ...where, project: { ...(where.project ?? {}), userId: ctx.user.id } };
    }
    return where;
  }
}
